src/pint/models/model_builder.py

```diff
@@ -2,317 +2,366 @@
+        # Parse all the useful lines
+        multi_line = Counter()
+        for l in interesting_lines(lines_of(parfile), comments=("#", "C ")):
             k = l.split()
-            if (
-                k[0] in param.keys()
-            ):  # repeat parameter TODO: add JUMP1 even there is only one
-                if k[0] in repeat_par.keys():
-                    repeat_par[k[0]] += 1
-                else:
-                    repeat_par[k[0]] = 2
-                param[k[0] + str(repeat_par[k[0]])] = k[1:]
+            param_inpar[k[0]] = k[1:]
+            # Handle the Mulit-tag lines
+            multi_line[k[0]] += 1
+            if k[0] in self.repeatable_param:
+                repeat_par[k[0]].append(k[1:])
             else:
-                param[k[0]] = k[1:]
-        self.param_inparF = param
-        for key in repeat_par.keys():
-            self.param_inparF[key + str(1)] = self.param_inparF.pop(key)
-        pfile.close()
-        return self.param_inparF
-
-    def get_all_categories(self,):
-        """Obtain a dictionary from category to a list of instances."""
-        comp_category = defaultdict(list)
-        for k, cp in Component.component_types.items():
-            comp_category[cp.category].append(cp())
-        return dict(comp_category)
-
-    def get_comp_from_parfile(self, parfile):
-        """Right now we only have one component on each category."""
-        params_inpar = self.preprocess_parfile(parfile)
-        for cat, cmps in self.get_all_categories().items():
-            selected_c = None
-            for cpi in cmps:
-                if cpi.component_special_params:
-                    if any(par in params_inpar for par in cpi.component_special_params):
-                        selected_c = cpi
-                        # Once have match, stop searching
-                        break
-                    else:
-                        continue
-                else:
-                    if cpi.is_in_parfile(params_inpar):
-                        selected_c = cpi
-            if selected_c is not None:
-                self.select_comp[cat] = selected_c
+                if multi_line[k[0]] > 1:
+                    log.info(
+                        "Lines with duplicate keys in par file:"
+                        " {} and {}".format(k[0], k[1:])
+                    )
+        return param_inpar, repeat_par
 
-    def sort_components(self, category_order=DEFAULT_ORDER):
-        """Sort the components into order.
+    def choose_model(self, param_inpar):
+        """Choose the model components based on the parfile.
 
-        Parameters
-        ----------
-        category_order: list, optional
-           The order for the order sensitive component categories.
+        Parameter
+        ---------
+        param_inpar: str
+            Dictionary of the unique parameters in .par file with the key is the
+        parfile line. Function `.parse_parfile` returns this dictionary.
```

Comment:
Please use Sphinx's syntax to refer to a function so that the documentation has a live link.